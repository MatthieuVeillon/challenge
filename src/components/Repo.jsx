import React, { Component } from "react";
import styled from "styled-components";
import moment from "moment";

// Styled components

const RepoTitle = styled.h3`font-weight: 200;`;

const UpdatedAt = styled.p`
  text-align: right;
  margin-right: 50px;
  font-size: 14px;
`;

const RepoWrapper = styled.div`margin-bottom: 30px;`;

class Repo extends Component {
  convertDateIntoTimeAgo = date => {
    return moment(date).fromNow();
  };

  displayRepo = () => {
    let counter = 0;
    return this.props.userRepos.map((repo, i) => {
      counter++;
      if (this.props.top) {
        if (counter <= this.props.top) {
          return (
            <RepoWrapper key={i}>
              <RepoTitle>
                <a target="_blank" href={repo.html_url}>
                  {repo.name}
                </a>
              </RepoTitle>
              <p>{repo.description}</p>
              <UpdatedAt>
                updated {this.convertDateIntoTimeAgo(repo.updated_at)}
              </UpdatedAt>
            </RepoWrapper>
          );
        }
      } else {
        return (
          <RepoWrapper key={i}>
            <RepoTitle>
              <a target="_blank" href={repo.html_url}>
                {repo.name}
              </a>
            </RepoTitle>
            <p>{repo.description}</p>
            <UpdatedAt>
              {this.convertDateIntoTimeAgo(repo.updated_at)}
            </UpdatedAt>
          </RepoWrapper>
        );
      }
    });
  };

  render() {
    return <div>{this.displayRepo()}</div>;
  }
}
export default Repo;
