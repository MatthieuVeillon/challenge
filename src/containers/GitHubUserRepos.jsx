import React, { Component } from "react";

import styled from "styled-components";

import UserSearch from "components/UserSearch.jsx";
import RepoNumber from "components/RepoNumber.jsx";
import UserFilter from "components/UserFilter.jsx";
import Repo from "components/Repo.jsx";

// Styled components declaration
const FlexContainer = styled.div`
  display: flex;
  padding: 5px;
  margin: 5px;
  border-top: 1px solid black;
  border-bottom: 1px solid black;
  align-items: center;
  justify-content: space-between;
`;
class GitHubUserRepos extends Component {
  constructor() {
    super();
    this.state = {
      username: "",
      userRepos: []
    };
  }

  // Somehow of a hack to be able to use snapshot with JESTand pass state mock as a prop
  componentWillMount() {
    this.setState({
      userRepos: this.props.userRepos || [],
      username: this.props.username || ""
    });
  }

  updateUserName = userInput => {
    if (userInput) {
      this.setState({ username: userInput }, this.fetchData);
    }
  };

  fetchData = () => {
    let url = `https://api.github.com/users/${this.state.username}/repos`;
    fetch(url)
      .then(response => response.json())
      .then(results => this.setState({ userRepos: results }))
      .then(() => this.filterByRecent());
  };

  updateUserRepos = fetchedData => {
    this.setState({ userRepos: fetchedData });
  };

  convertDatesIntoInt = dateStr => {
    return new Date(dateStr).getTime();
  };

  filterByRecent = () => {
    let repoSortedRecent = this.state.userRepos.sort(
      (a, b) =>
        this.convertDatesIntoInt(b.updated_at) -
        this.convertDatesIntoInt(a.updated_at)
    );
    this.setState({ userRepos: repoSortedRecent }, () =>
      console.log("Recent", this.state.userRepos)
    );
  };

  filterByOld = () => {
    let repoSortedOld = this.state.userRepos.sort(
      (a, b) =>
        this.convertDatesIntoInt(a.updated_at) -
        this.convertDatesIntoInt(b.updated_at)
    );
    this.setState({ userRepos: repoSortedOld });
  };

  render() {
    return (
      <div>
        {this.props.message}
        <UserSearch updateUserName={this.updateUserName} />
        {/* if there is a repo corresponding to the username entered by the user, display it or them  
        otherwise if there is no corresponding repo to the username entered, display a message telling the user, and finally 
        userinput is empty do not change the interface*/}
        {this.state.userRepos.length >= 1 ? (
          <FlexContainer>
            <RepoNumber
              username={this.state.username}
              userPicture={this.state.userRepos[0].owner.avatar_url}
              repoNumber={this.state.userRepos.length}
              repoNumberByPage={this.props.top}
            />
            <UserFilter
              filterByRecent={this.filterByRecent}
              filterByOld={this.filterByOld}
            />
          </FlexContainer>
        ) : this.state.userRepos.length === 0 && this.state.username ? (
          <p>No repositories found for {this.state.username}</p>
        ) : null}
        <Repo top={this.props.top} userRepos={this.state.userRepos} />
      </div>
    );
  }
}

export default GitHubUserRepos;
