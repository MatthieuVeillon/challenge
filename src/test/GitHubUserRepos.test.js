import React from "react";
import ReactDOM from "react-dom";
import GitHubUserRepos from "containers/GitHubUserRepos";
import Repo from "components/Repo.jsx";
import renderer from "react-test-renderer";

it("should render no repo by default", () => {
  const tree = renderer.create(<GitHubUserRepos />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("should render the repos from a user if a val username matchs a repo", () => {
  const userRepos = [
    {
      name: "repo1",
      description: "this is a repo",
      updated_at: "2017-11-27T17:09:38Z",
      html_url: "https://github.com/MatthieuVeillon/Alibay-project",
      owner: {
        avatar_url: "https://avatars1.githubusercontent.com/u/20015180?v=4"
      }
    },
    {
      name: "repo2",
      description: "this is a repo2",
      updated_at: "2017-11-27T17:09:38Z",
      html_url: "https://github.com/MatthieuVeillon/Alibay-project",
      owner: {
        avatar_url: "https://avatars1.githubusercontent.com/u/20015180?v=4"
      }
    }
  ];

  const tree = renderer.create(
    <GitHubUserRepos top={10} userRepos={userRepos} />
  );
  expect(tree).toMatchSnapshot();
});

it("should allow the user to sort the repo by the oldest", () => {
  const userRepos = [
    {
      name: "repo1",
      description: "this is a repo",
      updated_at: "2017-11-27T17:09:38Z",
      html_url: "https://github.com/MatthieuVeillon/Alibay-project",
      owner: {
        avatar_url: "https://avatars1.githubusercontent.com/u/20015180?v=4"
      }
    },
    {
      name: "repo2",
      description: "this is a repo2",
      updated_at: "2017-11-27T17:09:38Z",
      html_url: "https://github.com/MatthieuVeillon/Alibay-project",
      owner: {
        avatar_url: "https://avatars1.githubusercontent.com/u/20015180?v=4"
      }
    }
  ];

  const component = renderer.create(
    <GitHubUserRepos top={10} userRepos={userRepos} />
  );
  component.getInstance().filterByOld();

  expect(component.toJSON()).toMatchSnapshot();
});

it("should not display any repo if the userInput doesnt match any valid username", () => {
  const username = ";hkasdjkbfalskdjhuasdbfasdjgkagd";

  const tree = renderer.create(
    <GitHubUserRepos top={10} username={username} />
  );
  expect(tree).toMatchSnapshot();
});
