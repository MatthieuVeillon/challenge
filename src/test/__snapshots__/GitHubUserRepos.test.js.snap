// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should allow the user to sort the repo by the oldest 1`] = `
<div>
  <div
    className="sc-bdVaJa feMDxg"
  >
    <p>
      Github username
    </p>
    <form
      action="submit"
      onSubmit={[Function]}
    >
      <input
        placeholder="username"
        type="text"
      />
      <input
        type="submit"
        value="GO"
      />
    </form>
  </div>
  <div
    className="sc-bZQynM cunbwb"
  >
    <div
      className="sc-bwzfXH YDagW"
    >
      <img
        alt=""
        className="sc-htpNat oRhHv"
        src="https://avatars1.githubusercontent.com/u/20015180?v=4"
      />
      <div>
        <h4>
          Repositories for 
          
        </h4>
        <div
          className="sc-bwzfXH YDagW"
        >
          <p>
            Found 
            2
            .
          </p>
          <p>
            Showing top :
            10
          </p>
        </div>
      </div>
    </div>
    <div>
      <p>
        Filter by
      </p>
      <form
        action=""
      >
        <select
          onChange={[Function]}
        >
          <option
            value="recent"
          >
            Most recent first
          </option>
          <option
            value="oldest"
          >
            Oldest first
          </option>
        </select>
      </form>
    </div>
  </div>
  <div>
    <div
      className="sc-EHOje dgvmnA"
    >
      <h3
        className="sc-bxivhb gzkONY"
      >
        <a
          href="https://github.com/MatthieuVeillon/Alibay-project"
          target="_blank"
        >
          repo1
        </a>
      </h3>
      <p>
        this is a repo
      </p>
      <p
        className="sc-ifAKCX cGUfgB"
      >
        updated 
        18 days ago
      </p>
    </div>
    <div
      className="sc-EHOje dgvmnA"
    >
      <h3
        className="sc-bxivhb gzkONY"
      >
        <a
          href="https://github.com/MatthieuVeillon/Alibay-project"
          target="_blank"
        >
          repo2
        </a>
      </h3>
      <p>
        this is a repo2
      </p>
      <p
        className="sc-ifAKCX cGUfgB"
      >
        updated 
        18 days ago
      </p>
    </div>
  </div>
</div>
`;

exports[`should not display any repo if the userInput doesnt match any valid username 1`] = `
<div>
  <div
    className="sc-bdVaJa feMDxg"
  >
    <p>
      Github username
    </p>
    <form
      action="submit"
      onSubmit={[Function]}
    >
      <input
        placeholder="username"
        type="text"
      />
      <input
        type="submit"
        value="GO"
      />
    </form>
  </div>
  <p>
    No repositories found for 
    ;hkasdjkbfalskdjhuasdbfasdjgkagd
  </p>
  <div />
</div>
`;

exports[`should render no repo by default 1`] = `
<div>
  <div
    className="sc-bdVaJa feMDxg"
  >
    <p>
      Github username
    </p>
    <form
      action="submit"
      onSubmit={[Function]}
    >
      <input
        placeholder="username"
        type="text"
      />
      <input
        type="submit"
        value="GO"
      />
    </form>
  </div>
  <div />
</div>
`;

exports[`should render the repos from a user if a val username matchs a repo 1`] = `
<div>
  <div
    className="sc-bdVaJa feMDxg"
  >
    <p>
      Github username
    </p>
    <form
      action="submit"
      onSubmit={[Function]}
    >
      <input
        placeholder="username"
        type="text"
      />
      <input
        type="submit"
        value="GO"
      />
    </form>
  </div>
  <div
    className="sc-bZQynM cunbwb"
  >
    <div
      className="sc-bwzfXH YDagW"
    >
      <img
        alt=""
        className="sc-htpNat oRhHv"
        src="https://avatars1.githubusercontent.com/u/20015180?v=4"
      />
      <div>
        <h4>
          Repositories for 
          
        </h4>
        <div
          className="sc-bwzfXH YDagW"
        >
          <p>
            Found 
            2
            .
          </p>
          <p>
            Showing top :
            10
          </p>
        </div>
      </div>
    </div>
    <div>
      <p>
        Filter by
      </p>
      <form
        action=""
      >
        <select
          onChange={[Function]}
        >
          <option
            value="recent"
          >
            Most recent first
          </option>
          <option
            value="oldest"
          >
            Oldest first
          </option>
        </select>
      </form>
    </div>
  </div>
  <div>
    <div
      className="sc-EHOje dgvmnA"
    >
      <h3
        className="sc-bxivhb gzkONY"
      >
        <a
          href="https://github.com/MatthieuVeillon/Alibay-project"
          target="_blank"
        >
          repo1
        </a>
      </h3>
      <p>
        this is a repo
      </p>
      <p
        className="sc-ifAKCX cGUfgB"
      >
        updated 
        18 days ago
      </p>
    </div>
    <div
      className="sc-EHOje dgvmnA"
    >
      <h3
        className="sc-bxivhb gzkONY"
      >
        <a
          href="https://github.com/MatthieuVeillon/Alibay-project"
          target="_blank"
        >
          repo2
        </a>
      </h3>
      <p>
        this is a repo2
      </p>
      <p
        className="sc-ifAKCX cGUfgB"
      >
        updated 
        18 days ago
      </p>
    </div>
  </div>
</div>
`;
